File: src/main/java/demo/labrynth/LabyrinthCharacter.java
package demo.labrynth;

import com.jogamp.opengl.GL4;
import engine.feature.animation.BasicAnimation;
import engine.feature.collision.BoundingBox;
import engine.util.geometry.PointF;
import engine.feature.collision.collider.SpeculativeCollider;
import engine.core.ControllableObject;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.awt.event.KeyEvent;
import java.util.ArrayList;

//TODO: manage collisions with multiple objects at the same time
//TODO: cleanup code
public class LabyrinthCharacter extends ControllableObject implements SpeculativeCollider {

    private boolean[] keys;
    private ArrayList<PointF> collisionPoints;
    private BoundingBox nextBox;
    private BoundingBox currentBottomPlatform;

    private boolean isWalking;
    private boolean canJump;
    private float jumpTime;

    LabyrinthCharacter(int bufferParamsCount, int verticesCount, GL4 gl, float posX, float posY, Dimension boxDim, int id, float frameSizeX, float frameSizeY, BasicAnimation... animationSet) throws Exception {
        super(bufferParamsCount, verticesCount, gl, posX, posY, boxDim, id, frameSizeX, frameSizeY, animationSet);
        this.keys = new boolean[1000000];

        this.collisionPoints = new ArrayList<>();
        initCollisionPoints(collisionPoints);

        this.nextBox = new BoundingBox(posX, posY, getWidth(), getHeight());

        this.currentBottomPlatform = null;
        this.isWalking = false;
        this.canJump = true;
        this.jumpTime = 0f;
    }

    private void initCollisionPoints(ArrayList<PointF> target){
        target.add(new PointF(getPosX(), getPosY()));
        target.add(new PointF(getPosX(), getBottomY()));
        target.add(new PointF(getRightX(), getBottomY()));
        target.add(new PointF(getRightX(), getPosY()));
        target.add(new PointF(getPosX() + getWidth() / 2, getPosY()));
        target.add(new PointF(getPosX() + getWidth() / 2, getBottomY()));
        target.add(new PointF(getPosX(), getPosY() + getHeight() / 2));
        target.add(new PointF(getRightX(), getPosY() + getHeight() / 2));
        target.add(new PointF(getPosX(), getPosY() + getHeight() / 4)); // 8
        target.add(new PointF(getPosX(), getBottomY() - getHeight() / 4)); // 9
        target.add(new PointF(getRightX(), getPosY() + getHeight() / 4)); // 10
        target.add(new PointF(getRightX(), getBottomY() - getHeight() / 5)); // 11
    }

    //TODO: fix bug with horizontal collision
    @Override
    public void reactToCollision(@NotNull BoundingBox anotherBox) {
        if (detectBottomContact(anotherBox)) {
            setJumpState(false);
            currentBottomPlatform = anotherBox;
        }

        if (currentBottomPlatform != null && !detectBottomContact(currentBottomPlatform)) {
            currentBottomPlatform = null;
            setJumpState(true);
        }

        if (anotherBox.containsPoint(true, this.collisionPoints)) {
            processCollision(anotherBox);
        }
    }

    //TODO: fix bug - collision to horizontal above the vertical is ignored
    private void processCollision(BoundingBox anotherBox) {
        float moveX = this.getIntersectionWidth(anotherBox),
                moveY = this.getIntersectionHeight(anotherBox);
        boolean horizontalContact = detectHorizontalContact(anotherBox);
        boolean fallingState = !getJumpState() && this.getVelocityY() < 0f;

        //System.out.println("1 MoveX: " + moveX + ".  MoveY: " + moveY);

        if (horizontalContact)
            moveY = 0f;
        else
            moveX = 0f;

        //System.out.println("2 MoveX: " + moveX + ".  MoveY: " + moveY);

        if (this.getVelocityX() != 0f && this.getVelocityY() != 0f) {
            this.setVelocityX(0f);
            if (fallingState)
                this.setVelocityY(0f);
            this.setPosX(this.getPosX() + moveX);
            this.setPosY(this.getPosY() + moveY);
        } else if (this.getVelocityX() != 0f) {
            this.setPosX(this.getPosX() + moveX);
            this.setPosY(this.getPosY() + moveY);
            this.setVelocityX(0f);
        } else if (this.getVelocityY() != 0f) {
            this.setPosY(this.getPosY() + moveY);
            this.setPosX(this.getPosX() + moveX);
            this.setVelocityY(0f);
        }
    }

    private boolean detectBottomContact(BoundingBox anotherBox) {
        return anotherBox.containsNumberOfPoints(2, false,
                this.collisionPoints.get(1),
                this.collisionPoints.get(2),
                this.collisionPoints.get(5));
    }

    private boolean detectHorizontalContact(BoundingBox anotherBox) {
        return anotherBox.containsNumberOfPoints(2, true,
                this.collisionPoints.get(0),
                this.collisionPoints.get(1),
                this.collisionPoints.get(6)) ||
                anotherBox.containsNumberOfPoints(2, true,
                        this.collisionPoints.get(2),
                        this.collisionPoints.get(3),
                        this.collisionPoints.get(7)) ||
                anotherBox.containsAnyPointOf(true,
                        this.collisionPoints.get(8),
                        this.collisionPoints.get(9),
                        this.collisionPoints.get(10),
                        this.collisionPoints.get(11));
    }

    @Override
    public void update(float deltaTime) {

        applyVelocityX();
        applyVelocityY();

        processJumpTime(deltaTime);
        processGravityEffect(deltaTime);

        processAnimation();
        changePosition(deltaTime);

        updateCollisionPoints();
        updateNextBox(deltaTime);

        playAnimation(deltaTime);

        //System.out.println("Detect bottom contact " + (currentBottomPlatform == null ? "NULL" :
          //      detectBottomContact(currentBottomPlatform)) + "posY " + posY);
        //System.out.println(currentBottomPlatform == null ? "NULL" : currentBottomPlatform.toString());
        //System.out.println("JUMPSTATE:" + jumpState);
        //System.out.println(this.velocityX + " " + this.velocityY);

    }

    private void applyVelocityX() {
        if (keys[KeyEvent.VK_D]) {
            isWalking = true;
            this.setVelocityX(3.5f);
        } else if (keys[KeyEvent.VK_A]) {
            isWalking = true;
            this.setVelocityX(-3.5f);
        } else
            this.setVelocityX(0f);
    }

    private void applyVelocityY() {
        if (keys[KeyEvent.VK_S]) {
            this.setVelocityY(5.0f);
            this.setJumpState(true);
            this.canJump = false;
        } else if (keys[KeyEvent.VK_W] && canJump)
            jump();
        else if (!getJumpState())
            this.setVelocityY(0.0f);
    }

    private void processJumpTime(float deltaTime) {
        if (!canJump) {
            this.jumpTime += deltaTime;
            float jumpTimeLimit = 600f;
            if (jumpTime >= jumpTimeLimit) {
                jumpTime = 0f;
                canJump = true;
            }
        }
    }

    private void processGravityEffect(float deltaTime) {
        float gravity = 1f;
        if (getJumpState() && currentBottomPlatform == null) {
            //System.out.println("Applying gravity");
            this.setVelocityY(this.getVelocityY() + (gravity * deltaTime) / 10);
        }
    }

    private void processAnimation() {
        if (this.getVelocityX() == 0 && this.getVelocityY() == 0) {
            this.setAnimation(this.getAnimations().get(2));
            this.getCurrentAnim().setCurrentFrameX(0);
            this.getCurrentAnim().setCurrentFrameY(2);
            this.isWalking = false;
        } else if (this.getVelocityY() != 0 && this.getVelocityX() != 0) {

        } else if (this.getVelocityX() != 0 && !getJumpState()) {
            if (getCurrentAnim() != getAnimations().get(0))
                this.setWalkAnim();
        } else if (this.getVelocityX() != 0 && getJumpState()) {
            this.setJumpAnimation();
        } else if (this.getVelocityY() > 0f && this.getVelocityY() <= 0.2f) {
            this.setAnimation(this.getAnimations().get(2));
            this.getCurrentAnim().setCurrentFrameX(0);
            this.getCurrentAnim().setCurrentFrameY(2);
        }

    }

    private void changePosition(float deltaTime) {
        this.setPosY(this.getPosY() + (this.getVelocityY() * deltaTime) / 20);
        this.setPosX(this.getPosX() + (this.getVelocityX() * deltaTime) / 20);
    }

    private void updateCollisionPoints() {
        collisionPoints.set(0, new PointF(getPosX(), getPosY()));
        collisionPoints.set(1, new PointF(getPosX(), getBottomY()));
        collisionPoints.set(2, new PointF(getRightX(), getBottomY()));
        collisionPoints.set(3, new PointF(getRightX(), getPosY()));
        collisionPoints.set(4, new PointF(getPosX() + getWidth() / 2, getPosY()));
        collisionPoints.set(5, new PointF(getPosX() + getWidth() / 2, getBottomY()));
        collisionPoints.set(6, new PointF(getPosX(), getPosY() + getHeight() / 2));
        collisionPoints.set(7, new PointF(getRightX(), getPosY() + getHeight() / 2));
    }

    @Override
    public void preventCollision() {
        this.setVelocityY(0.0f);
        this.setVelocityX(0.0f);
        keys[KeyEvent.VK_W] = false;
        keys[KeyEvent.VK_S] = false;
        keys[KeyEvent.VK_A] = false;
        keys[KeyEvent.VK_D] = false;
        isWalking = false;
    }

    private void updateNextBox(float deltaTime) {
        nextBox.setPosition(getPosX() + getVelocityX() * deltaTime, getPosY() + getVelocityY() * deltaTime);
    }

    @NotNull
    @Override
    public BoundingBox getNextBox() {
        return this.nextBox;
    }

    @Override
    public void keyTyped(@NotNull KeyEvent e) {
    }

    @Override
    public void keyPressed(@NotNull KeyEvent e) {

        if (!this.keys[e.getKeyCode()])
            this.keys[e.getKeyCode()] = true;

        if (e.getKeyCode() == KeyEvent.VK_D || e.getKeyCode() == KeyEvent.VK_A) {
            if (!isWalking && !getJumpState()) {
                setWalkAnim();
            }
        } else if (e.getKeyCode() == KeyEvent.VK_W)
            if (!getJumpState())
                setJumpAnimation();

    }

    @Override
    public void keyReleased(@NotNull KeyEvent e) {

        if (this.keys[e.getKeyCode()])
            this.keys[e.getKeyCode()] = false;

        switch (e.getKeyCode()) {
            case KeyEvent.VK_D:
                break;
            case KeyEvent.VK_A:
                break;
            case KeyEvent.VK_W:
                break;
            case KeyEvent.VK_S:
                break;
            default:
                break;
        }

    }

    private void jump() {
        setVelocityY(getVelocityY() - 25f);
        setAirFloating();
    }

    private void setAirFloating() {
        setJumpState(true);
        canJump = false;
    }

    private void setJumpAnimation() {
        setCurrentAnim(getAnimations().get(1));
        getCurrentAnim().setCurrentFrameY(1);
        getCurrentAnim().setCurrentFrameX(7);
        getCurrentAnim().setFirstPosX(7);
        getCurrentAnim().setLastPosX(9);
    }

    private void setWalkAnim() {
        setCurrentAnim(getAnimations().get(0));
        getCurrentAnim().setCurrentFrameY(2);
        getCurrentAnim().setCurrentFrameX(1);
        getCurrentAnim().setFirstPosX(1);
        getCurrentAnim().setLastPosX(6);
    }

}

File: src/main/java/demo/labrynth/LabyrinthCharacter.kt
package demo.labrynth

import engine.core.*
import engine.feature.collision.BoundingBox
import engine.util.geometry.PointF
import java.awt.event.KeyEvent
import java.util.*

class LabyrinthCharacter internal constructor(
        posX: Float,
        posY: Float,
        width: Float,
        height: Float,
        animatedObject: AnimatedObject,
        graphicalObject: OpenGlObject2D
) : CompositeObject(animatedObject, null, graphicalObject) {

    private val keyboardItems: BooleanArray = BooleanArray(1000000)
    private val collisionPoints: ArrayList<PointF> = arrayListOf()

    private var currentBottomPlatform: BoundingBox? = null
    private var isWalking: Boolean = true
    private var canJump: Boolean = true
    private var jumpTime: Float = 0f

    private val xVelocityModifier = 3.5f
    private val yVelocityModifier = 5f
    private val jumpTimeLimit = 600f
    private val deltaGravityModifier = 10
    private val deltaModifier = 20

    private val controllableObject: ControllableObject = object : ControllableObject() {

        override fun reactToCollision(anotherBox: BoundingBox) {
            if (isBottomContact(anotherBox)) {
                jumpState = false
                currentBottomPlatform = anotherBox
            }
            if (currentBottomPlatform != null && !isBottomContact(currentBottomPlatform!!)) {
                currentBottomPlatform = null
                jumpState = true
            }
            if (anotherBox.containsPoint(true, collisionPoints)) {
                processCollision(anotherBox)
            }
        }

        override fun keyTyped(e: KeyEvent) {}

        override fun keyPressed(e: KeyEvent) {
            if (!keyboardItems[e.keyCode]) keyboardItems[e.keyCode] = true
            if (e.keyCode == KeyEvent.VK_D || e.keyCode == KeyEvent.VK_A) {
                if (!isWalking && !jumpState) {
                    setWalkAnim()
                }
            } else if (e.keyCode == KeyEvent.VK_W) if (!jumpState) setJumpAnim()
        }

        override fun keyReleased(e: KeyEvent) {
            if (keyboardItems[e.keyCode]) keyboardItems[e.keyCode] = false
        }
    }

    init {
        setCtrlComponent(controllableObject)
        initCollisionPoints(collisionPoints)

        val box = BoundingBox(posX, posY, width, height)
        setBoundingBox(box)

        currentBottomPlatform = null
        isWalking = false
        canJump = true
        jumpTime = 0f
    }

    private fun initCollisionPoints(target: ArrayList<PointF>) {
        with(graphicalComponent.box) {
            this?.let {
                target.add(PointF(it.posX, it.posY)) // 0
                target.add(PointF(it.posX, it.bottomY)) // 1
                target.add(PointF(it.rightX, it.bottomY)) // 2
                target.add(PointF(it.rightX, it.posY)) // 3
                target.add(PointF(it.posX + it.width / 2, it.posY)) // 4
                target.add(PointF(it.posX + it.width / 2, it.bottomY)) // 5
                target.add(PointF(it.posX, it.posY + it.height / 2)) // 6
                target.add(PointF(it.rightX, it.posY + it.height / 2)) // 7
                target.add(PointF(it.posX, it.posY + it.height / 4)) // 8
                target.add(PointF(it.posX, it.bottomY - it.height / 4)) // 9
                target.add(PointF(it.rightX, it.posY + it.height / 4)) // 10
                target.add(PointF(it.rightX, it.bottomY - it.height / 5)) // 11
                target.add(PointF(it.posX, it.posY + it.height / 3)) // 12
                target.add(PointF(it.rightX, it.posY + it.height / 3)) // 13
                target.add(PointF(it.posX, it.bottomY - it.height / 3)) // 14
                target.add(PointF(it.rightX, it.bottomY - it.height / 3)) // 15
            }
        }
    }

    private fun processCollision(anotherBox: BoundingBox) {
        var moveX: Float = graphicalComponent.box!!.getIntersectionWidth(anotherBox)
        var moveY: Float = graphicalComponent.box!!.getIntersectionHeight(anotherBox)
        val horizontalContact = detectHorizontalContact(anotherBox)
        val fallingState = !jumpState && velocityY < 0f

        if (horizontalContact) {
            moveY = 0f
        } else {
            moveX = 0f
        }

        if (velocityX != 0f && velocityY != 0f) {
            velocityX = 0f
            if (fallingState) {
                velocityY = 0f
            }
        } else if (velocityX != 0f) {
            velocityX = 0f
        } else if (velocityY != 0f) {
            velocityY = 0f
        }

        graphicalComponent.box!!.posX += moveX
        graphicalComponent.box!!.posY += moveY
    }

    private fun isBottomContact(anotherBox: BoundingBox): Boolean {
        return anotherBox.containsNumberOfPoints(2, false,
                collisionPoints[1],
                collisionPoints[2],
                collisionPoints[5])
    }

    private fun detectHorizontalContact(anotherBox: BoundingBox): Boolean {
        return anotherBox.containsNumberOfPoints(2, true,
                collisionPoints[0],
                collisionPoints[1],
                collisionPoints[6]) ||
                anotherBox.containsNumberOfPoints(2, true,
                        collisionPoints[2],
                        collisionPoints[3],
                        collisionPoints[7]) ||
                anotherBox.containsAnyPointOf(true,
                        collisionPoints[8],
                        collisionPoints[9],
                        collisionPoints[10],
                        collisionPoints[11],
                        collisionPoints[12],
                        collisionPoints[13])
    }

    override fun update(deltaTime: Float) {
        applyVelocityX()
        applyVelocityY()
        processJumpTime(deltaTime)
        processGravityEffect(deltaTime)
        processAnimation()
        changePosition(deltaTime)
        updateCollisionPoints()
        animationComponent!!.playAnimation(deltaTime)
    }

    override fun react(entity: Entity) {
        if (entity is OpenGlObject2D) {
            val box = entity.box
            box?.let { controlComponent?.reactToCollision(it) }
        }
    }

    private fun applyVelocityX() {
        when {
            keyboardItems[KeyEvent.VK_D] -> {
                isWalking = true
                velocityX = xVelocityModifier
            }
            keyboardItems[KeyEvent.VK_A] -> {
                isWalking = true
                velocityX = -xVelocityModifier
            }
            else -> velocityX = 0f
        }
    }

    private fun applyVelocityY() {
        if (keyboardItems[KeyEvent.VK_S]) {
            velocityY = yVelocityModifier
            jumpState = true
            canJump = false
        } else if (keyboardItems[KeyEvent.VK_W] && canJump) {
            jump()
        } else if (!jumpState) {
            velocityY = 0.0f
        }
    }

    private fun processJumpTime(deltaTime: Float) {
        if (!canJump) {
            jumpTime += deltaTime
            val limit = jumpTimeLimit
            if (jumpTime >= limit) {
                jumpTime = 0f
                canJump = true
            }
        }
    }

    private fun processGravityEffect(deltaTime: Float) {
        val gravity = 1f
        if (jumpState && currentBottomPlatform == null) { //System.out.println("Applying gravity");
            velocityY += gravity * deltaTime / deltaGravityModifier
        }
    }

    private fun processAnimation() = with(animationComponent) {
        this?.let {
            val isIdleAnim = velocityX == 0f && velocityY == 0f
            val isWalkAnim = velocityX != 0f && !jumpState && it.currentAnim != it.animations[0]
            val isJumpAnim = velocityX != 0f && jumpState
            val isInsignificantJump = velocityY > 0f && velocityY <= 0.2f
            val isDiagonalMove = velocityX != 0f && velocityY != 0f

            when {
                isIdleAnim -> setIdleAnim()
                isDiagonalMove -> Unit
                isWalkAnim -> setWalkAnim()
                isJumpAnim -> setJumpAnim()
                isInsignificantJump -> setWalkAnim()
                else -> Unit
            }
        }
    }

    private fun changePosition(deltaTime: Float) = with(graphicalComponent.box) {
        this?.let {
            it.posY += velocityY * deltaTime / deltaModifier
            it.posX += velocityX * deltaTime / deltaModifier
        }
    }

    private fun updateCollisionPoints() = with(graphicalComponent.box) {
        this?.let {
            collisionPoints[0] = PointF(it.posX, it.posY)
            collisionPoints[1] = PointF(it.posX, it.bottomY)
            collisionPoints[2] = PointF(it.rightX, it.bottomY)
            collisionPoints[3] = PointF(it.rightX, it.posY)
            collisionPoints[4] = PointF(it.posX + it.width / 2, it.posY)
            collisionPoints[5] = PointF(it.posX + it.width / 2, it.bottomY)
            collisionPoints[6] = PointF(it.posX, it.posY + it.height / 2)
            collisionPoints[7] = PointF(it.rightX, it.posY + it.height / 2)
            collisionPoints[8] = PointF(it.posX, it.posY + it.height / 4)
            collisionPoints[9] = PointF(it.posX, it.bottomY - it.height / 4)
            collisionPoints[10] = PointF(it.rightX, it.posY + it.height / 4)
            collisionPoints[11] = PointF(it.rightX, it.bottomY - it.height / 5)
            collisionPoints[12] = PointF(it.posX, it.posY + it.height / 3)
            collisionPoints[13] = PointF(it.rightX, it.posY + it.height / 3)
            collisionPoints[14] = PointF(it.rightX, it.bottomY - it.height / 3)
            collisionPoints[15] = PointF(it.rightX, it.bottomY - it.height / 3)
        }
    }

    private fun jump() {
        velocityY -= 20f
        setAirFloating()
    }

    private fun setAirFloating() {
        jumpState = true
        canJump = false
    }

    private fun setIdleAnim() = animationComponent?.let {
        it.setAnimation(it.animations[2])
        it.currentAnim.setCurrentFrameX(0)
        it.currentAnim.setCurrentFrameY(2)
        isWalking = false
    }

    private fun setJumpAnim() = animationComponent?.let {
        it.currentAnim = it.animations[1]
        it.currentAnim.currentFrameY = 1
        it.currentAnim.currentFrameX = 7
        it.currentAnim.setFirstPosX(7)
        it.currentAnim.setLastPosX(9)
    }

    private fun setWalkAnim() = animationComponent?.let {
        it.currentAnim = it.animations[0]
        it.currentAnim.currentFrameY = 2
        it.currentAnim.currentFrameX = 1
        it.currentAnim.setFirstPosX(1)
        it.currentAnim.setLastPosX(6)
    }
}